ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/AttestationUtil.java

```diff
@@ -137,10 +163,104 @@ public static Bytes32 getAttestationMessageToSign(AttestationData attestationDat
+    }
+    return returnValue;
+  }
+
+  public static void validate_indexed_attestation(BeaconState state, IndexedAttestation indexed_attestation){
+    //Verify validity of ``indexed_attestation``.
+    List<UnsignedLong> bit_0_indices = indexed_attestation.getCustody_bit_0_indices();
+    List<UnsignedLong> bit_1_indices = indexed_attestation.getCustody_bit_1_indices();
+
+    //Verify no index has custody bit equal to 1 [to be removed in phase 1]
+    checkArgument((bit_0_indices.size() == 0);
+    //Verify max number of indices
+    checkArgument(bit_0_indices.size() + bit_1_indices.size() <= MAX_INDICES_PER_ATTESTATION);
+    //Verify index sets are disjoint
+    checkArgument(intersection(bit_0_indices, bit_1_indices).size() == 0);
+    //Verify indices are sorted
+    Collections.sort(bit_0_indices);
+    List<UnsignedLong> bit_0_indices_sorted = new ArrayList<UnsignedLong>(bit_0_indices);
+    Collections.sort(bit_0_indices_sorted);
+    List<UnsignedLong> bit_1_indices_sorted = new ArrayList<UnsignedLong>(bit_1_indices);
+    Collections.sort(bit_1_indices_sorted);
+    checkArgument(bit_0_indices.equals(bit_0_indices_sorted) && bit_1_indices.equals(bit_1_indices_sorted));
+
+    List<Validator> validators = state.getValidator_registry();
+    List<BLSPublicKey> pubkeys = Arrays.stream(validators.subList(0, bit_0_indices.size()).toArray()).map((validator) -> {return ((Validator)validator).getPubkey();}).collect(Collectors.toList());
+    pubkeys.addAll(Arrays.stream(validators.subList(0, bit_1_indices.size()).toArray()).map((validator) -> {return ((Validator)validator).getPubkey();}).collect(Collectors.toList()));
+
+    List<Bytes32> message_hashes = new ArrayList<Bytes32>();
+    message_hashes.add(hash_tree_root(LIST_OF_BASIC, new AttestationDataAndCustodyBit(indexed_attestation.getData(), false).toBytes()));
+    message_hashes.add(hash_tree_root(LIST_OF_BASIC, new AttestationDataAndCustodyBit(indexed_attestation.getData(), true).toBytes()));
+
+    BLSSignature signature = indexed_attestation.getSignature();
+    long domain = get_domain(state, DOMAIN_ATTESTATION, indexed_attestation.getData().getTarget_epoch());
+    //Verify aggregate signature
+    checkArgument(BLSVerify.bls_verify_multiple(pubkeys, message_hashes, signature, UnsignedLong.valueOf(domain)));
+  }
+
+  public static UnsignedLong get_attestation_data_slot(BeaconState state, AttestationData data) {
+    UnsignedLong committee_count = get_epoch_committee_count(state, data.getTarget_epoch());
+    UnsignedLong offset = (data.getCrosslink().getShard().plus(UnsignedLong.valueOf(SHARD_COUNT)).minus(get_epoch_start_shard(state, data.getTarget_epoch()))).mod(UnsignedLong.valueOf(SHARD_COUNT));
+    return UnsignedLong.valueOf(Math.floorDiv(get_epoch_start_slot(data.getTarget_epoch()).plus(offset).longValue(), Math.floorDiv(committee_count.longValue(), (long) SLOTS_PER_EPOCH)));
+  }
+
+  public static <T> List<T> intersection(List<T> list1, List<T> list2) {
```

Comment:
This can be optimized using a TreeSet I think. Can you add a TODO here that says optimize?

Reply:
One-liner for the intersection of two lists:
`List<T> intersection = list1.stream().filter(list2::contains).collect(Collectors.toSet());`

If order doesn't matter then use Collectors.toSet(); otherwise, Collectors.toList()